MapboxNavigation/InstructionPresenter.swift

```diff
@@ -14,26 +14,63 @@ class InstructionPresenter {
     var onShieldDownload: ShieldDownloadCompletion?
 
     private let imageRepository = ImageRepository.shared
-
+    
     func attributedText() -> NSAttributedString {
-        guard let label = self.label else {
-            return NSAttributedString()
-        }
-
         let string = NSMutableAttributedString()
-
-        for component in instruction {
+        fittedAttributedComponents().forEach { string.append($0) }
+        return string
+    }
+    
+    func fittedAttributedComponents() -> [NSAttributedString] {
+        guard let label = self.label else { return [] }
+        var attributedPairs = self.attributedPairs()
+        let availableBounds = label.availableBounds()
+        let totalWidth = attributedPairs.attributedStrings.map { $0.size() }.reduce(.zero, +).width
+        let stringFits = totalWidth <= availableBounds.width
+        
+        guard !stringFits else { return attributedPairs.attributedStrings }
+        
+        let indexedComponents = attributedPairs.components.enumerated().map { IndexedVisualInstructionComponent(component: $1, index: $0) }
+        let filtered = indexedComponents.filter { $0.component.abbreviation != nil }
+        let sorted = filtered.sorted { $0.component.abbreviationPriority < $1.component.abbreviationPriority }
+        for component in sorted {
+            let isFirst = component.index == 0
+            let joinChar = isFirst ? "" : " "
+            guard component.component.type == .text else { continue }
+            guard let abbreviation = component.component.abbreviation else { continue }
+            
+            attributedPairs.attributedStrings[component.index] = NSAttributedString(string: joinChar + abbreviation, attributes: attributesForLabel(label))
```

Comment:
Would it make sense to check we're within bounds here? I'm just slightly weary of bracket lookups like this and how fragile they are.

Reply:
I agree about the fragility, but this could cover up an unwanted bug where we could abbreviate
"West Fremont Avenue" as "West W Ave" or crash due to out of bounds.

So instead, I added an assertion to make sure we would catch such a bug early in https://github.com/mapbox/mapbox-navigation-ios/pull/1169/commits/a8d8fafc0a6a0ed8f7aa1ab882972996d9c6e6a8